File: src/api-client.ts
/**
 * xAI Grok API クライアント（X検索用）
 */

/** カンマ区切りのXハンドル文字列をリストに変換する。@/＠プレフィックスを除去。 */
export function parseHandles(rawValue: string): string[] {
  if (!rawValue) return [];
  const handles: string[] = [];
  for (const part of rawValue.replace(/\uff20/g, "@").split(",")) {
    let handle = part.trim();
    if (!handle) continue;
    if (handle.startsWith("@")) handle = handle.slice(1);
    if (handle) handles.push(handle);
  }
  return handles;
}

/** YYYY-MM-DD形式の日付文字列を検証する。空文字はnullを返す。 */
export function validateIsoDate(value: string, label: string): string | null {
  if (!value) return null;
  if (!/^\d{4}-\d{2}-\d{2}/.test(value) || isNaN(new Date(value).getTime())) {
    throw new Error(`${label}はYYYY-MM-DD形式で指定してください: ${value}`);
  }
  return value;
}

/** Grok API レスポンスからテキストを抽出する。 */
export function extractTextFromResponse(payload: Record<string, unknown>): string {
  // 最優先: トップレベル output_text
  const outputText = payload.output_text;
  if (outputText) {
    if (Array.isArray(outputText)) {
      return outputText.map(String).join("\n").trim();
    }
    return String(outputText).trim();
  }

  // output 配列からメッセージを抽出
  let outputs = payload.output;
  if (!Array.isArray(outputs)) outputs = [];

  const texts: string[] = [];
  for (const item of outputs as unknown[]) {
    if (typeof item === "string") {
      texts.push(item);
      continue;
    }
    if (typeof item !== "object" || item === null) continue;
    const obj = item as Record<string, unknown>;
    if (obj.type !== "message") continue;

    const content = obj.content;
    if (typeof content === "string") {
      if (content) texts.push(content);
      continue;
    }
    if (!Array.isArray(content)) continue;

    for (const block of content) {
      if (typeof block === "string") {
        if (block) texts.push(block);
        continue;
      }
      if (typeof block !== "object" || block === null) continue;
      const b = block as Record<string, unknown>;
      if (b.type === "output_text" || b.type === "text") {
        const textValue = (b.text || b.output_text) as string | undefined;
        if (textValue) texts.push(String(textValue));
      }
    }
  }
  if (texts.length > 0) return texts.join("\n").trim();

  // フォールバック: JSON全体
  return JSON.stringify(payload);
}

export interface SearchOptions {
  query: string;
  systemPrompt: string;
  maxResults?: number;
  allowedHandles?: string[] | null;
  excludedHandles?: string[] | null;
  fromDate?: string | null;
  toDate?: string | null;
  freshness?: string;
  searchMode?: string;
  language?: string;
  enableImageUnderstanding?: boolean;
  enableVideoUnderstanding?: boolean;
  temperature?: number;
  maxOutputTokens?: number;
  timeoutSeconds?: number;
}

export class XSearchAPIClient {
  private apiKey: string;
  private apiBase: string;
  private model: string;

  constructor() {
    this.apiKey = process.env.XAI_API_KEY || process.env.GROK_API_KEY || "";
    this.apiBase = process.env.XAI_API_BASE || "https://api.x.ai/v1";
    this.model = process.env.XAI_GROK_MODEL || "grok-4-0709";
  }

  async search(options: SearchOptions): Promise<string> {
    if (!this.apiKey) {
      return "XAI_API_KEY (またはGROK_API_KEY) が設定されていません。";
    }

    const toolConfig: Record<string, unknown> = {
      max_results: options.maxResults ?? 8,
      search_mode: options.searchMode ?? "auto",
      freshness: options.freshness ?? "auto",
      language: options.language ?? "ja",
    };
    if (options.allowedHandles?.length)
      toolConfig.allowed_x_handles = options.allowedHandles;
    if (options.excludedHandles?.length)
      toolConfig.excluded_x_handles = options.excludedHandles;
    if (options.fromDate) toolConfig.from_date = options.fromDate;
    if (options.toDate) toolConfig.to_date = options.toDate;
    if (options.enableImageUnderstanding)
      toolConfig.enable_image_understanding = true;
    if (options.enableVideoUnderstanding)
      toolConfig.enable_video_understanding = true;

    const payload = {
      model: this.model,
      input: [
        { role: "system", content: options.systemPrompt },
        { role: "user", content: options.query.trim() },
      ],
      tools: [{ type: "x_search", x_search: toolConfig }],
      temperature: options.temperature ?? 0.2,
      max_output_tokens: options.maxOutputTokens ?? 900,
    };

    const url = `${this.apiBase.replace(/\/+$/, "")}/responses`;
    const timeoutMs = (options.timeoutSeconds ?? 45) * 1000;

    let response: Response;
    try {
      response = await fetch(url, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(timeoutMs),
      });
    } catch (err) {
      return `Grok APIへの接続に失敗しました: ${err}`;
    }

    // body は一度しか読めないので text() で取得してから JSON.parse
    const rawBody = await response.text();

    if (response.status >= 300) {
      try {
        const errorPayload = JSON.parse(rawBody);
        const err = errorPayload.error;
        let errorMessage: string;
        if (typeof err === "object" && err !== null) {
          errorMessage = err.message || JSON.stringify(err);
        } else if (typeof err === "string" && err) {
          errorMessage = err;
        } else {
          errorMessage = JSON.stringify(errorPayload);
        }
        return `Grok APIエラー(${response.status}): ${errorMessage}`;
      } catch {
        return `Grok APIエラー(${response.status}): ${rawBody}`;
      }
    }

    let responsePayload: Record<string, unknown>;
    try {
      responsePayload = JSON.parse(rawBody);
    } catch {
      return `Grok APIの応答を解析できませんでした: ${rawBody}`;
    }

    const text = extractTextFromResponse(responsePayload);
    return text || JSON.stringify(responsePayload);
  }
}

File: x_search_mcp/api_client.py
"""xAI Grok API 非同期クライアント（X検索用）"""

from __future__ import annotations

import json
import logging
import os
from datetime import datetime
from typing import List, Optional

import httpx

logger = logging.getLogger("x-search-mcp")


def parse_handles(raw_value: str) -> List[str]:
    """カンマ区切りのXハンドル文字列をリストに変換する。

    '@'プレフィックスと全角'＠'を正規化して除去する。
    """
    if not raw_value:
        return []
    handles = []
    for part in raw_value.replace("＠", "@").split(","):
        handle = part.strip()
        if not handle:
            continue
        if handle.startswith("@"):
            handle = handle[1:]
        if handle:
            handles.append(handle)
    return handles


def validate_iso_date(value: str, label: str) -> Optional[str]:
    """YYYY-MM-DD形式の日付文字列を検証する。"""
    if not value:
        return None
    try:
        datetime.fromisoformat(value)
    except ValueError as exc:
        raise ValueError(
            f"{label}はYYYY-MM-DD形式で指定してください: {value}"
        ) from exc
    return value


def extract_text_from_response(payload: dict) -> str:
    """Grok API レスポンスからテキストを抽出する。

    payload.output_text → output[].content[].text → JSON全体 の順に試行する。
    """
    # 最優先: トップレベル output_text（Responses API の標準フィールド）
    if payload.get("output_text"):
        if isinstance(payload["output_text"], list):
            return "\n".join(str(t) for t in payload["output_text"]).strip()
        return str(payload["output_text"]).strip()

    # output 配列からメッセージを抽出
    outputs = payload.get("output", [])
    if not isinstance(outputs, list):
        outputs = []

    texts: List[str] = []
    for item in outputs:
        if isinstance(item, str):
            texts.append(item)
            continue
        if not isinstance(item, dict):
            continue
        if item.get("type") != "message":
            continue
        content = item.get("content", [])
        # content が文字列の場合
        if isinstance(content, str):
            if content:
                texts.append(content)
            continue
        # content が配列の場合
        if not isinstance(content, list):
            continue
        for block in content:
            if isinstance(block, str):
                if block:
                    texts.append(block)
                continue
            if not isinstance(block, dict):
                continue
            block_type = block.get("type", "")
            if block_type in ("output_text", "text"):
                text_value = block.get("text") or block.get("output_text")
                if text_value:
                    texts.append(str(text_value))
    if texts:
        return "\n".join(texts).strip()

    return json.dumps(payload, ensure_ascii=False)


class XSearchAPIClient:
    """xAI Grok API を使用した X 検索クライアント"""

    def __init__(self) -> None:
        self.api_key: str = os.getenv("XAI_API_KEY") or os.getenv("GROK_API_KEY", "")
        self.api_base: str = os.getenv("XAI_API_BASE", "https://api.x.ai/v1")
        self.model: str = os.getenv("XAI_GROK_MODEL", "grok-4-0709")
        self._client: Optional[httpx.AsyncClient] = None

    @property
    def headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    async def _get_client(self) -> httpx.AsyncClient:
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                headers=self.headers,
                timeout=60.0,
            )
        return self._client

    async def close(self) -> None:
        if self._client and not self._client.is_closed:
            await self._client.aclose()

    async def search(
        self,
        query: str,
        *,
        system_prompt: str,
        max_results: int = 8,
        allowed_handles: Optional[List[str]] = None,
        excluded_handles: Optional[List[str]] = None,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
        freshness: str = "auto",
        search_mode: str = "auto",
        language: str = "ja",
        enable_image_understanding: bool = False,
        enable_video_understanding: bool = False,
        temperature: float = 0.2,
        max_output_tokens: int = 900,
        timeout_seconds: int = 45,
    ) -> str:
        """Grok API の x_search ツールを使用して X を検索する。

        Returns:
            検索結果のテキスト、またはエラーメッセージ。
        """
        if not self.api_key:
            return "XAI_API_KEY (またはGROK_API_KEY) が設定されていません。"

        tool_config: dict = {
            "max_results": max_results,
            "search_mode": search_mode,
            "freshness": freshness,
            "language": language,
        }
        if allowed_handles:
            tool_config["allowed_x_handles"] = allowed_handles
        if excluded_handles:
            tool_config["excluded_x_handles"] = excluded_handles
        if from_date:
            tool_config["from_date"] = from_date
        if to_date:
            tool_config["to_date"] = to_date
        if enable_image_understanding:
            tool_config["enable_image_understanding"] = True
        if enable_video_understanding:
            tool_config["enable_video_understanding"] = True

        tool_entry = {"type": "x_search", "x_search": tool_config}

        payload = {
            "model": self.model,
            "input": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": query.strip()},
            ],
            "tools": [tool_entry],
            "temperature": temperature,
            "max_output_tokens": max_output_tokens,
        }

        url = f"{self.api_base.rstrip('/')}/responses"

        try:
            client = await self._get_client()
            response = await client.post(
                url, json=payload, timeout=timeout_seconds
            )
        except httpx.RequestError as exc:
            return f"Grok APIへの接続に失敗しました: {exc}"

        if response.status_code >= 300:
            try:
                error_payload = response.json()
                err = error_payload.get("error", "")
                if isinstance(err, dict):
                    error_message = err.get("message", "") or json.dumps(err, ensure_ascii=False)
                elif isinstance(err, str) and err:
                    error_message = err
                else:
                    error_message = json.dumps(error_payload, ensure_ascii=False)
            except Exception:
                error_message = response.text
            return f"Grok APIエラー({response.status_code}): {error_message}"

        try:
            response_payload = response.json()
        except ValueError:
            return f"Grok APIの応答を解析できませんでした: {response.text}"

        text = extract_text_from_response(response_payload)
        if not text:
            text = json.dumps(response_payload, ensure_ascii=False)

        return text
